### PR TITLE
Fix pagination limit

### PR DESCRIPTION
**Pull request by rajpnz**
New PR that targets develop. 

This is a work around for the 100 project pagination limit in dependency track.
The solution was suggested by the developer of the framework.
https://groups.io/g/dependency-track/topic/71890367?p=Created,,,20,1,0,0::recentpostdate/sticky,,,20,2,0,71890367

### CODE REVIEW

**Comment by pmckeown** [comment]
This still needs a test for a few reasons:

1. We had a defect and the first thing to do is to create a test to describe the defect which then passes when the fix is implemented.  This documents the limitation on the server for future devs and proves that we are now not affected.
2. As the maintainer of this repo, I have no way to know that the fix you are proposing is fixed without cloning your repo and either adding a test myself or testing it manually by scripting the insert of > 100 projects into a Dependency Track instance.  Either way you are not making it easy for the maintainer to accept your change which in the OSS community is not likely to make a maintainer accept your PR.
3. Regression protection from future changes or bugs in the use of those query parameters

**Comment by rajpnz** [comment]
> This still needs a test for a few reasons:
> 
> 1. We had a defect and the first thing to do is to create a test to describe the defect which then passes when the fix is implemented.  This documents the limitation on the server for future devs and proves that we are now not affected.
> 2. As the maintainer of this repo, I have no way to know that the fix you are proposing is fixed without cloning your repo and either adding a test myself or testing it manually by scripting the insert of > 100 projects into a Dependency Track instance.  Either way you are not making it easy for the maintainer to accept your change which in the OSS community is not likely to make a maintainer accept your PR.
> 3. Regression protection from future changes or bugs in the use of those query parameters

Ok, I'll add a test for this

File: src/test/java/io/github/pmckeown/dependencytrack/TestResourceConstants.java [hunk 1]
```diff
@@ -9,6 +9,8 @@ private TestResourceConstants() {
     public static final String V1_METRICS_PROJECT_CURRENT = "/api/v1/metrics/project/(.*)/current";
     public static final String V1_METRICS_PROJECT_REFRESH = "/api/v1/metrics/project/(.*)/refresh";
     public static final String V1_PROJECT_UUID = "/api/v1/project/(.*)";
+    public static final String V1_PROJECT_WITH_PAGINATION = "/api/v1/project";
```
Comment (pmckeown) [inline src/test/java/io/github/pmckeown/dependencytrack/TestResourceConstants.java:12]: The prod code no longer makes a call without the pagination parameters so why would the test code need to?

File: src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		CommonConfig commonConfig = new CommonConfig();
+		commonConfig.setDependencyTrackBaseUrl("http://localhost:" + wireMockRule.port());
+		projectClient = new ProjectClient(commonConfig);
+	}
+
+	@Test
+	public void thatCallingGetProjectsCallsApiWithoutPaginationAndReturnsAllProjects() {
+		stubFor(get(urlEqualTo(V1_PROJECT_WITHOUT_PAGINATION)).willReturn(
+				aResponse().withBodyFile("api/v1/project/get-all-projects.json")));
+		stubFor(get(urlEqualTo(V1_PROJECT_WITH_PAGINATION)).willReturn(
```
Comment (pmckeown) [inline src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java:41]: This stubbing in unnecessary as a request without the query parameters is not being made.

File: src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import io.github.pmckeown.dependencytrack.AbstractDependencyTrackMojoTest;
+import io.github.pmckeown.dependencytrack.CommonConfig;
+
+public class ProjectClientIntegrationTest extends AbstractDependencyTrackMojoTest {
```
Comment (pmckeown) [inline src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java:24]: I don't see how this test is different from the existing integration tests that stub out a call to get projects and verifies the response contains the expected amount.

File: src/test/resources/__files/api/v1/project/get-all-projects-with-pagination.json [hunk 4]
```diff
@@ -0,0 +1,123 @@
+[
```
Comment (pmckeown) [inline src/test/resources/__files/api/v1/project/get-all-projects-with-pagination.json:1]: Is this just a copy of the `get-all-projects.json` file?  If so what is the benefit of duplicating it?

**Comment by rajpnz** [comment]
> I was hoping to see a test that expressed the defect and the fix. This change doesn't do that, merely duplicates existing tests and stubs out calls that are not made by the production code.
> 
> Its ugly but a JSON file that contains 101 projects could be created and then stubbed for return in a specific test that describes the defect. I did not get round to creating builder classes for the response building which makes the testing harder here as a massive JSON file is needed instead of a succinct call to something like `ResponseBuilder.aResponseWithProjects(101).build()`. But happy to have a massive JSON file that I can address when I add the Response Builder classes.
> 
> The test could be something like:
> 
> ```
> @Test
> public void thatWhenMoreProjectsExistThanTheServerDefaultPageSizeTheResponseIsNotPaginated() {
>     // Configure wiremock with the appropriate response
>     // Verify count of projects is greater than the Server Default page size of 100
> }
> ```

I was thinking of creating a json file with a 101 projects, but as you said it was a bit ugly. Also in future
dependency track may change its pagination limit and 101 will not be valid. To me the essence of this test is that we call the api with a limit query param that is high enough to by pass the pagination issue. 

This was another way I was going to write the test. 
```
@Test
public void thatCallingGetProjectsCallsApiWithHighLimitAndReturnsAllProjects() {

	stubFor(get(urlPathEqualTo(V1_PROJECT))
			.withQueryParam("limit", matching(HIGH_LIMIT_REGEX))
			.withQueryParam("offset", equalTo("0"))
			.willReturn(aResponse().withBodyFile("api/v1/project/get-all-projects.json")));

	List<Project> projects = projectClient.getProjects().getBody().orElse(Collections.emptyList());

	assertThat(projects.size(), is(COUNT_ALL_PROJECTS));
	verify(exactly(1),
			getRequestedFor(urlPathEqualTo(V1_PROJECT))
					.withQueryParam("limit", matching(HIGH_LIMIT_REGEX))
					.withQueryParam("offset", equalTo("0")));

}
```

File: src/test/java/io/github/pmckeown/dependencytrack/TestResourceConstants.java (hunk 1 above)
Comment (rajpnz) [reply src/test/java/io/github/pmckeown/dependencytrack/TestResourceConstants.java:12]: In the test I wanted to express that if we call the GET projects api with just "/api/v1/project" we expect a different set of projects returned. 
In saying that I'm comfortable in dropping this and the associated stub, as the test will still fail if someone changes back to using "/api/v1/project".

File: src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java (hunk 2 above)
Comment (rajpnz) [reply src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java:41]: see above comment about dropping this.

File: src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java (hunk 3 above)
Comment (rajpnz) [reply src/test/java/io/github/pmckeown/dependencytrack/project/ProjectClientIntegrationTest.java:24]: Yes I agree it similar. 
The other integration tests were focused on the testing the mojos, and I would have had to piggy back on one or all of those that called ProjectClient.getProjects().  
I felt it was better to create a test solely focused on ProjectClient.

File: src/test/resources/__files/api/v1/project/get-all-projects-with-pagination.json (hunk 4 above)
Comment (rajpnz) [reply src/test/resources/__files/api/v1/project/get-all-projects-with-pagination.json:1]: It's got one less project so if some calls the api just "/api/v1/project", the count of projects returned will be different and the test will fail.

**Pull request merged**